# የአማርኛ ጽሁፎችን መተንተን

የአማርኛ ጽሁፎችን ከድረ-ገጽ ወይም ከበፋይል በመውሰድ ትንተና ልናደርግበት እንፈልግ ይሆናል። ከመሰረታዊ ትንተና እስከ ላቀ ትንተና ልናደርግ እንችላለን። መሰረታዊ ትንተና የምንላቸው እንደምሳሌ በጽሁፉ ውስጥ ስንት አረፍተነገራት አሉ፤ ወይም ስንት ቃላት ይገኛሉ። በብዛት የተደጋገመው ቃል የትኛው ነው።  የላቀ ትንተና ልንላቸው የምንችላቸው ለምሳሌ ትርጉም ሊሰጡ የሚችሉ ጥንድ ቃላትን ከ ሰንዱ መለየት፣ ሰነዱ ስለምን እንደሚያወራና የሰነዱን ስሜት (ደስታ፣ ሃዘን፣ ንዴት እና ስላቅ....) መለየት መቻል፣ ሰነዱን ገጽታውን ሳያዛቡ በተፈለገው ርዝማኔ አሳጥሮ ማቅረብ። 

በዚህ ክፍል የምንመለከተው የጽሁፍ ምንጫችን ፋይል ይሆናል። ከ አማርኛ ድረ-ገጾች ላይ መረጃዎችን መሰብሰብ በሌላ ክፍል የምንመለከተው ይሆናል።    

## ፋይልን ለትንተና መክፈት እና ይዘቱን ማንበብ

ብዙ የፋይል አይነቶች አሉ። በዚህ ክፍል የምንመለከተው ዝርግ የጽሁፍ (ፍላት ቴክስት) ፋይልን ነው። ለምናደርገው  ትንተናዎች ይረዳ ዘንድ፣ የምንጠቀመው የአቤ ጉበኛ መጽሀፍ የሆነውን አልወለድምን ነው። መጽሐፉን ውሂብ በሚለው አቃፊ ውስጥ ታገኙታላችሁ። 

ትንተና ከማድረጋችን በፊት ፋይልን መጀመሪያ መክፈት አለብን። በፓይተን ብዙ የተለያዩ ፋይልን የመክፈቻ መንገዶች አሉ። የመጀመሪያው ዘዴ ፋይሉን ተለዋዋጭ በመጠቀም መክፈት፤ ተለዋዋጩን በመጠቀም የተከፈተውን ጽሁፍ ማንበብ። ሁለተኛው ዘዴ መክፈቱን እና ማንበቡን አንድ ላይ አጣምሮ ማካሄድ። ሁለቱንም በተናጠል እንመለከታቸዋለን። 

*** 
መሰረታዊ የሆኑት ፋይል ላይ ልናደርጋቸው የምንችላቸው ተግባራት:
* __open()__ :- ፋይሉን ለመክፈት የሚጠቅመን ነው። ለመክፈት የፈለግነውን ፋይል ስም ተጠቅመን መክፈት እንችላለን። 
* __read()__ :- የፋይሉን ይዘት ባጠቃላይ አንዴ ለማንበብ። ትላልቅ ፋይሎችን በምንከፍት ጊዜ ሜሞሪ ሊፈልግ ይችላል። 
* __readline()__ :- የፋይሉን ይዘት መስመር በመስመር ለማንበብ ይጠቅማል። ብዙ ሜሞሪ አይፈጅም። 
* __write()__ :- ሕብረቁምፊ (ስትሪንግ) ወደ ፋይሉ ለመጻፍ። 
* __close()__ :- የከፈትነውን ፋይል ለመዝጋት። 
*** 



In [3]:
# የመጀመሪያውን ዘዴ በመጠቀም ፋይሉን መክፈት።
# ፋይሉን ዶሴ በሚለው ተለዋዋጭ ላይ መክፈት 
ዶሴ = open('./ውሂብ/አልወለድም_በአቤ_ጉበኛ.txt', 'r')
# በመቀጠልም ዶሴን ተጠቅመን የፋይሉን ይዞታ ማንበብ 
# readline() መጠቀም እንችላለን 
ዶሴ.read()
ዶሴ.close()

In [5]:
# ሁለተኛውን ዘዴ በመጠቀም ፋይሉን መክፈት። 
# ማለትም ማንበቡን እና መክፈቱን አንድ ላይ ማከናወን። 
# readline() መጠቀም እንችላለን 
with open('./ውሂብ/አልወለድም_በአቤ_ጉበኛ.txt', 'r') as ዶሴ:
    ፋይል=ዶሴ.read()

የመረጣችሁትን መንገድ መጠቀም ትችላላችሁ። እኔ ሁለተኛውን መንገድ እጠቀማለሁ። አሁን ፋይሉ ባጠቃላይ ፋይል በሚለው ተለዋዋጭ ላይ ተነቦ እናገኘዋለን። ለምሳሌ የመጀመሪያ 20 ፊደላት እንደሚከተለው መመልከት እንችላለን። ይህም ፋይሉ በትክክል መነበቡን ለመገንዘብ ይረዳናን። 

In [12]:
# የመጀመሪያ 20 ፊደላትን እንመልከት 
# ፋይል[0:20] መጻፍ እንችላለን። 
ፋይል[:20]

'\ufeffአዲስ መግቢያ ።\n\n\tይህ መጽሐ'

## ቅድመ ትንተናዊ ማስተካከያ

ከውጤቱ እንደምንመለከተው ፋይሉ ተነቧል። ነገር ግን \ufeff n\n\t የሚሉ የአማርኛ ያልሆኑ ፊደላት ተቀላቅለውበታል። የመጀመሪያው `\ufeff` ባይት ኦርደር ማርክ ወይም ባጭሩ ስለፋይሉ ምንነት ለሚያነበው ፕሮግራም ምልክት ሰጪ ነው። በኛ ፋይል መጀመሪያ ላይ መገኘቱ ጽሁፋችን ዩኒኮድ መሆኑን የሚጠቁም ሲሆን ለፕሮግራማችን ጽሁፉ ዩኒኮድ እናም የባይት ኦርደር ማርክ ምልክት አለው ብለን ከነገርነው ይሄንን ምልክት በራሱ ያስወግደዋል። ይህንንም ለማድረግ ክፈት በሚለው ተግባር ውስጥ `encoding='utf-8-sig'` ማስቀመጥ ነው። ይሄም የዩቲኤፍ ስምንት (የዩኒኮድ) መግለጫው ሲሆን ሲግ ደግሞ ፊርማ ወይም መታወቂያ እንደማለት ነው። የዩኒኮድ መታወቂያ መጀመሪያው ላይ አለ ማለት ነው። ፓይተን ይሄንን ሲያይ `\ufeff` ን ከማንበብ ይገደባል ማለት ነው።    

In [13]:
# የባይት ኦርደር ማርክ እንዳለው ለፓይተን አሳውቀን ፋይሉን እንክፈት 
# የመጀመሪያ 20 ውን ፊደላት እንዳለፈው እንመልከት ፋይል[0:20]
with open('./ውሂብ/አልወለድም_በአቤ_ጉበኛ.txt', 'r', encoding='utf-8-sig') as ዶሴ:
    ፋይል=ዶሴ.read()
ፋይል[0:20]

'አዲስ መግቢያ ።\n\n\tይህ መጽሐፍ'

ከውጤቱ የምናስተውለው ሌላኛው ችግር \n እና \t የሚሉ የአማርኛ ያልሆኑ ምልክቶች ናቸው። እነዚህም የአዲስ መስመር መጀመሪያ እና የትር (ታብ) የምንለው አመላካች ናቸው። የአዲስ መስመር ምልክቱን `ፋይል=ዶሴ.read()` የሚለውን መስመር `ፋይል=ዶሴ.read().replace('\n', '')` በሚለው ብንተካው `\n` ከጽሁፉ ውስጥ ይቀንስልናል። `replace('\n', '')` የሚለው የአዲስ መስመር ምልክት የሆነውን `'\n'` በ ባዶ ቦታ ተካው ማለት ነው።  

In [14]:
# የአዲስ መስመር ምልክት የሆነው '\n' በ ቦዶ ቦታ replace('\n', '') በመጠቀም መተካት
# የመጀመሪያ 20 ውን ፊደላት እንዳለፈው እንመልከት ፋይል[0:20]
with open('./ውሂብ/አልወለድም_በአቤ_ጉበኛ.txt', 'r', encoding='utf-8-sig') as ዶሴ:
    ፋይል=ዶሴ.read().replace('\n', '')
ፋይል[0:20]

'አዲስ መግቢያ ።\tይህ መጽሐፍ የ'

በመጨረሻ የሚቀረን የትር ምልከት የሆነው `\t` ነው። እሱንም `replace()` እንደሚለው ተግባር ሁሉ ፓይተን `maketrans()` የሚል የሕብረቁምፊ ዘዴ አለው። ማለትም `maketrans('አ', 'በ')` ብንል አን በ በ ተካልን ማለት ነው። ስለዚህም  `maketrans('\t', '')` ብንጠቀም ትርን ከ ሕብረቁምፊው ይቀንስልናል ማለት ነው። በምሳሌ እንመልከት፡

In [19]:
# የ maketrans() ዘዴ(ሜትድ) ምሳሌ 
ሕብረ_ቁምፊ ='\tይህ መጽሐፍ'
# ትር ከመቀነሱ በፊት 
ሕብረ_ቁምፊ01 = ሕብረ_ቁምፊ
ሕብረ_ቁምፊ01 
ሰንጠረዥ = str.maketrans("\t", (" "))
# ትር ከተቀነሰ በኋላ
ሕብረ_ቁምፊ= ሕብረ_ቁምፊ.translate(ሰንጠረዥ)
ሕብረ_ቁምፊ01, ሕብረ_ቁምፊ

('\tይህ መጽሐፍ', ' ይህ መጽሐፍ')

ይሔንን መንገድ በመከተል `ፋይል` የሚለውን ተለዋዋጭ ከላይ የተሞከርነውን ተጠቅመን `\t` ከጽሁፉ መቀነስ እንችላለን ማለት ነው። 

In [22]:
# ትርን ከጽሁፋችን ለመቀነስ የተከተልነው መንገድ። 
# የመጀመሪያ 20 ውን ፊደላት እንዳለፈው እንመልከት ፋይል[0:20]
with open('./ውሂብ/አልወለድም_በአቤ_ጉበኛ.txt', 'r', encoding='utf-8-sig') as ዶሴ:
    ፋይል=ዶሴ.read().replace('\n', '')

# ትርን በባዶ ቦታ መተካት 
ሰንጠረዥ = str.maketrans("\t", (" "))
# ሰንጠረዥን በፋይ ላይ መተግበር
ፋይል= ፋይል.translate(ሰንጠረዥ)
ፋይል[0:20]

'አዲስ መግቢያ ። ይህ መጽሐፍ የ'

## የቃላት እና የአረፍተነገራት ብዛት በፋይሉ

ቃላት በአማርኛ ጽሁፍ አንዳንዴ በ ሁለት ነጥብ አንዳንዴ ባዶ ቦታን በመተው ይጻፋሉ። አረፍተነገራት በአብዛኛው በ አራት ነጥብ ያበቃሉ። የተለያዩ ምልክቶችም የአረፍተነገራትን መጨረሻ ሊያመለክቱ ይችላሉ። ስለዚህም እነዚህን በሚገባ ለይቶ ማወቅና ወደፕሮግራም መለወጥ ያስፈልጋል። ይህ ጽሁፍ መነሻ እንደመሆኑ መጠን የሚገተሉትን ግምቶች ታሳቢ ያደርጋል። 
* አረፍተ ነገራት በ አራት ነጥብ __።__ ያበቃሉ በሚል ታሳቢ አድርጎ ይነሳል። 
* የቃላት መለያ ባዶ ቦታ ነው ብሎ ታሳቢ ያደርጋል 

በእነዚህ ሁለቱን ታሳቢ በማድረግ የአረፍተነግር እና የቃላት ብዛትን በ ፋይሉ ውስጥ እንደሚከተለው ተሰርተዋል።  

In [34]:
# የቃላት እና የአረፍተነገራት ብዛት በ አልወለድም ጽሁፍ ውስጥ 
with open('./ውሂብ/አልወለድም_በአቤ_ጉበኛ.txt', 'r', encoding='utf-8-sig') as ዶሴ:
    ፋይል=ዶሴ.read().replace('\n', '')

    # ትርን በባዶ ቦታ መተካት 
    ሰንጠረዥ = str.maketrans("\t", (" "))
    # ሰንጠረዥን በፋይ ላይ መተግበር
    ፋይል= ፋይል.translate(ሰንጠረዥ)
    # የቃላት እና የ በአራት ነጥብ የሚያልቁ አረፍተነገራት ብዛት 
    print('የቃላት ብዛት:\t\t       ', len(ፋይል.split()))
    # ይሄንን ስክሪፕት በመጠቀም ማረጋገጥ ይቻላል 
    # grep '።' -o አልወለድም_በአቤ_ጉበኛ.txt | wc -l
    print('አረፍተነገር ብዛት (በ \'።\' የሚጨርሱ):    ', ፋይል.count('።'))

የቃላት ብዛት:		        31842
አረፍተነገር ብዛት (በ '።' የሚጨርሱ):     1720


***
ከላይ ያሉት ውጤቶች ተቀራራቢ ግምት ናቸው እንጂ የታወቀ ትክክለኛ መልስ ነው ማለት አይደለም። ለምሳሌ ከላይ የተጠቀምነው `ፋይል.split()` የሚሰራው ባዶ ቦታ ሲያገኝ መቁጠር ነው። በጽሁፍ ስህተት ምክኒያት ቃላት በመካከላቸው ሁለት ነጥብ ኖር ነገር ግን ባዶ ቦታ ከሌለ እንደ አንድ ቃል ይቆጥራቸዋል ማለት ነው። ስለዚህ ሁል ግዜም እንዲህ አይነት ትንተናዎች ሲሰሩ ያሉትን የተለያዩ አማራጮች ፕሮግራማችን እንዲያካትት ማድረግ ያስፈልጋል። 

## የቃላት ድግግሞሽ በፋይሉ

የቃላት ድግግሞሽ (ወርድ ፍርክዌንሲ) የምንለው በጽሁፉ ውስጥ እያንዳንድ ቃል ምን ያህል ግዜ እንደተደገመ ለማወቅ ይረዳናል። ብዙም የተለየ መረጃ ላይሰጠን ይችላል፤ ነገር ግን ምን ያህል ልዩ ቃላቶች በጽሁፉ ውስት ጥቅም ላይ እንደዋሉ፤ በብዛት የተደጋገሙ ቃላት የትኞቹ እንደሆኑ ማወቅ እንችላለን። በመቀጠልም በምን ያህል ርቀጥ እነዚህ ቃላት እንደተደጋገሙ መለየት እንችላለን። ይህም የሰነድ ጠቋሚ (ኢንዴክሰር) ለመጻፍ እንደ አንድ ግብአት ያገለግላል።   

*** 
የተለያዩ መንገዶችን በመጠቀም የቃላት ድግግሞሽን መስራት ቢቻልም በዚህ ጽሁፍ ሁለቱን መንገዶች እንመለከታለን። 
*** 

In [37]:
# መጀመሪያ ፋይሉን እንክፈትና እና ቃላቱን በዝርዝር ውስጥ እናስቀምጥ 

with open('./ውሂብ/አልወለድም_በአቤ_ጉበኛ.txt', 'r', encoding='utf-8-sig') as ዶሴ:
    ፋይል=ዶሴ.read().replace('\n', '')

    # ትርን በባዶ ቦታ መተካት (ሌሎችን እንጨምራለን እንደ «»!፣፧፡ ምልክቶች በጽሁፉ ውስጥ የሚገኙ
    # ነገር ግን ቃላት ያልሆኑ ናቸው። ) 
    # 8* ያደረግንበት በግራ በኩል እና በቀኝ በኩል ያሉት ርዝመታቸው እኩል መሆን ስላለበት ነው። 
    ሰንጠረዥ = str.maketrans("\t«»!፣፧፡።", 8*(" "))
    # ሰንጠረዥን በፋይ ላይ መተግበር
    ፋይል= ፋይል.translate(ሰንጠረዥ)

    # የቃላት ዝርዝር 
    የቃላት_ዝርዝር = [ቃል for ቃል in ፋይል.split()]

In [40]:
# ከቃላት ዝርዝሩ 10 ቃላቶችን ብንመለከት 
የቃላት_ዝርዝር[:10]

['አዲስ', 'መግቢያ', 'ይህ', 'መጽሐፍ', 'የዛሬ', '01', 'ዓመት', 'ገደማ', 'በደንቡ', 'ምርመራ']

### ዘዴ አንድ

In [45]:
# የቃላቱን ድግግሞሽ ለማውጣት ተመሳሳይ የሆኑ ቃላት ሲገኙ ድምራቸውን መጨመር ነው። 
# ይሄንንን ለማድረግ ተስማሚ የሆነው የውሂብ መዋቅር መዝገበ-ቃል ወይም ዲክሽነሪ ነው። 
# ቃሉን እንደቁልፍ (ኪይ) በመጠቀም ድግግሞሹን እንደ እሴት (ቫሊዩ) መጠቀም እንችላለን ማለት ነው። 
with open('./ውሂብ/አልወለድም_በአቤ_ጉበኛ.txt', 'r', encoding='utf-8-sig') as ዶሴ:
    ፋይል=ዶሴ.read().replace('\n', '')
    ሰንጠረዥ = str.maketrans("\t«»!፣፧፡።", 8*(" "))
    # ሰንጠረዥን በፋይ ላይ መተግበር
    ፋይል= ፋይል.translate(ሰንጠረዥ)

    # የቃላት ዝርዝር 
    የቃላት_ዝርዝር = [ቃል for ቃል in ፋይል.split()]
መዝገበ_ቃል = {} # ባዶ መዝገበ ቃል አወጅን 
# የ ፎር ድግግሞሽን በመጠቀም መዝገበ_ቃሉን በ ቃል እና ድግግሞሹ መሙላት እንችላለን። 
for ቃል in የቃላት_ዝርዝር:
    መዝገበ_ቃል[ቃል] +=1 # የቃላት_ዝርዝር[ቃል] እሴቱን ያመለክታል፣ ቃል ደግሞ ቁልፉን። 
    
    """ነገር ግን ይሄ ፕሮግራም አይሰራም። ምክኒያቱም መዝገበ_ቃሉ ባዶ ስለሆነ የመጀመሪያውን 
    ቁልፍ በሚያጣበት ግዜ  የቁክፍሳንካ (KeyError) ያመጣል ይህ እንዳይሆን መፍትሔውን በሚቀጥለው 
    ተመልከቱ"""

KeyError: 'አዲስ'

In [46]:
# ከላይ ለተፈጠረው ሳንካ መፍትሔ 
with open('./ውሂብ/አልወለድም_በአቤ_ጉበኛ.txt', 'r', encoding='utf-8-sig') as ዶሴ:
    ፋይል=ዶሴ.read().replace('\n', '')
    ሰንጠረዥ = str.maketrans("\t«»!፣፧፡።", 8*(" "))
    # ሰንጠረዥን በፋይ ላይ መተግበር
    ፋይል= ፋይል.translate(ሰንጠረዥ)

    # የቃላት ዝርዝር 
    የቃላት_ዝርዝር = [ቃል for ቃል in ፋይል.split()]
መዝገበ_ቃል = {} # ባዶ መዝገበ ቃል አወጅን 
# የ ፎር ድግግሞሽን በመጠቀም መዝገበ_ቃሉን በ ቃል እና ድግግሞሹ መሙላት እንችላለን። 
for ቃል in የቃላት_ዝርዝር:
    try:
        መዝገበ_ቃል[ቃል] +=1 # የቃላት_ዝርዝር[ቃል] እሴቱን ያመለክታል፣ ቃል ደግሞ ቁልፉን። 
    except KeyError:
        መዝገበ_ቃል[ቃል] = 1    

`መዝገበ_ቃል` ብለን በመጻፍ ቃላት እና ድግግሞሻቸውን ማየት እንችላለን ነገር ግን ይሄ ትልቅ ከሆነ የማንፈልገውን ቃል ሁሉ ይዘረግፍብናል። ስለዚህም ወይ የምንፈልገውን ቃል ከመዝገበ_ቃል መፈለግ፤ አልያም የራሳችንን ተግባር በመጻፍ የምንፈልገውን ያህል ከ መዝገበ_ቃል እየቀነሱ ማየት ይቻላል። 

In [47]:
# ለምሳሌ ነው 10 ሩን የቃላት ዝርዝር ስንመለከት የቃላት_ዝርዝር[:10] መጽሐፍ የሚል ቃል አለ። 
# መጽሐፍ ምን ያህል ግዜ በ አቤ ጉበኛ አልወለድም እንደተደጋገመ ለማየት 
መዝገበ_ቃል['መጽሐፍ']

8

የ መጽሐፍ ድግግሞሽ በጽሁፉ ውስጥ ከውጤቱ እንደምናየው 8 ጊዜ ተደጋግሟል ማለት ነው። መዝገበ ቃሉን ከብዙ ድግግሞሽ ጀምሮ እየቀነሰ እንዲሔድ ተራ ማስያዝ እንችላለን። የፈለግነውን ብዛት እየቀነስን ለመመልከት እንዲረዳን ነው።   

In [53]:
ተራ_የያዘ_መቃል = sorted(መዝገበ_ቃል.items(), key=lambda ቁልፍእሴት: ቁልፍእሴት[1], reverse=True)

In [50]:
# የፈለግነውን ያህል እየቀነስን ለማየት የሚያስችለንንን ተግባር እንጻፍ። 
from itertools import islice

def ውሰድ(ስንት, ድረስ):
    """ ከላይ የፈለግነውን ያህል ለመውሰድ ይረዳናል """
    return list(islice(ድረስ, ስንት))

In [54]:
# ውሰድን ተጠቅመን ከመዝገበ_ቃል 10 ቃላትን እንመልከት 
ውሰድ(10, ተራ_የያዘ_መቃል)

[('ነው', 306),
 ('ሁሉ', 202),
 ('ነበር', 172),
 ('ነገር', 172),
 ('ላይ', 171),
 ('ግን', 166),
 ('እንጂ', 149),
 ('ሰው', 142),
 ('ጊዜ', 137),
 ('ሰዎች', 137)]

### ዘዴ ሁለት

ፓይተን ቅድም ያነሳነውን በመዝገበ-ቃል ውስጥ ቁልፍ ያለመኖር ችግርን ለመፍታት ነባሪ መዝገበቃል አዘጋጅቷል። ይሄንን የፓይተን ቤተመዝገብ በመጠቀም የቃላት ድግግሞሽን መስራት እንችላለን። ቁልፍ ያለመኖር ችግርን ለማብራራት የሚከተሉት ፕሮግራሞች እንመልከት። 

In [57]:
# ቁልፍ ያለመኖር ችግርን ለማሳየት የገባ ምሳሌ፤ 
# ነባሪ መዝገበቃሉ እንዴት መፍትሔ እንደሰጠ እንመልከት

ሳንካ_ያለበት_መቃ = {}

#ሳንካ የሌለበት መዝገበ ቃል 
from collections import defaultdict
ሳንካ_የሌለበት_መቃ = defaultdict(int)

In [58]:
# ሳንካ ያለበትን መዝገበ ቃል ብንጠራው እንደሚከተለው ይሰራል። 
ሳንካ_ያለበት_መቃ

{}

In [59]:
# ሳንካ ያለበትን መዝገበ ቃል የሌለውን ቁልፍ ፈልግ ብንለው የቁልፍ ሳንካ ያሳየናል። 
ሳንካ_ያለበት_መቃ['ነው']

KeyError: 'ነው'

In [60]:
ሳንካ_የሌለበት_መቃ

defaultdict(int, {})

In [61]:
ሳንካ_የሌለበት_መቃ['ነው'] # በ ቁጥር ስላስነሳነው ለሌለ ቁልፍ ሁሉ ዜሮን ያስቀምጣል። 

0

In [65]:
## መዝገበቃሉን ማወጅ 
from collections import defaultdict
ሳንካ_የሌለበት_መቃ = defaultdict(int)
# ሁለተኛው ዘዴ  
with open('./ውሂብ/አልወለድም_በአቤ_ጉበኛ.txt', 'r', encoding='utf-8-sig') as ዶሴ:
    ፋይል=ዶሴ.read().replace('\n', '')
    ሰንጠረዥ = str.maketrans("\t«»!፣፧፡።", 8*(" "))
    # ሰንጠረዥን በፋይል ላይ መተግበር
    ፋይል= ፋይል.translate(ሰንጠረዥ)

    # የቃላት ዝርዝር 
    የቃላት_ዝርዝር = [ቃል for ቃል in ፋይል.split()]
    
    for ቃል in የቃላት_ዝርዝር:
        ሳንካ_የሌለበት_መቃ[ቃል] += 1

In [66]:
ሳንካ_የሌለበት_መቃ['ነው']

306

## የቃላት ቦታን መጠቆም

የቃላት ቦታን መጠቆም ስንል ለምሳሌ `ነው` በ ጽሁፉ ውስጥ 306 ጊዜ ተደጋግሟል። ነገር ግን የት የት ቦታ ላይ እንደተደጋገመ ቦታውን መጠቆም ማለት ነው። የቃሉን ስብጥር በጽሁፉ ውስጥ ለማየት ይጠቅማል።  

In [ ]:
#ስራችንን ለማከናወን enumerateን እንጠቀማለን።  
# ኢኑምሬት አንድ ነገር ወስዶ ሁለት አድርጎ የሚመልስ የፓይተን ተግባር ነው። 
list(enumerate('ገበገታ')) # የፊደላቱን ቦታ እና ፊደላቱን ይሰጠናል። 

In [69]:
## መዝገበ ቃል እንጠቀማለን። ቁጥር ሳይሆን ስብስብ (ሴት) ነው ብለን አውጀን እንነሳ። 
from collections import defaultdict
የቃላት_ቦታ_በሰነዱ = defaultdict(set)

with open('./ውሂብ/አልወለድም_በአቤ_ጉበኛ.txt', 'r', encoding='utf-8-sig') as ዶሴ:
    ፋይል=ዶሴ.read().replace('\n', '')
    ሰንጠረዥ = str.maketrans("\t«»!፣፧፡።", 8*(" "))
    # ሰንጠረዥን በፋይል ላይ መተግበር
    ፋይል= ፋይል.translate(ሰንጠረዥ)

    # ቃላቶች
    ቃላቶች = ፋይል.split()
    # enumerate() 
    for ጠቋሚ,ቃል in enumerate(ቃላቶች): 
        የቃላት_ቦታ_በሰነዱ[ቃል].add(ጠቋሚ)

In [70]:
# መጽሐፍ ብዛቱ 8 ሲሆን የት የት ቦታ እንደሚገኝ እንመልከት 
የቃላት_ቦታ_በሰነዱ['መጽሐፍ']

{3, 120, 418, 2044, 2092, 16299, 18929, 25814}

በዚህ ውጤት መሰረት የመጀመሪያው መጽሐፍ የሚለው መጽሐፍ የሚገኘው በአራተኛ ስፍራ ላይ ነው።  አራተኛ ያልንበት ምክኒያት ጠቋሚዎቹ መቁጠር የሚጀምሩት ከ ዜሮ ስለሆነ ነው።  ይኼንን `ፋይል[:20]` በመጻፍ ቃላቶቹን በመቁጠር ማረጋገጥ ይቻላል። በመቀጠል በ 121፣ 419፣ 2045፣ 2093፣ 16300፣ 18939 እና 25815ኛ ቦታዎች ላይ ይገኛል ማለት ነው። የሌሎቹንም ቃላቶች ቦታቸውን ለማየት `የቃላት_ቦታ_በሰነዱ['ቃል']` እያስገቡ መመልከት ይቻላል። 

In [72]:
ፋይል[:20]

'አዲስ መግቢያ   ይህ መጽሐፍ የ'

## ማጠቃለያ

እንዳይረዝም ይኼንን ክፍል በዚህ እናብቃና በሚቀጥለው ክፍል የቃላትን ቦታ በፋይ ውስጥ የሚጠቁመውን ፕሮግራም አንድ ደረጃ ከፍ በማድረግ የቃላትን ቦታ በአረፍተነገር ጠቋሚነት እንዴት መግለጽ እንደምንችል እንመለከታለን። ማለትም ለምሳሌ መጽሐፍ የሚለው ቃል 8 ግዜ ተደጋግሟል። ድግግሞሹ የተካሔደባቸው ስንት አረፍተነገራት እንደሆኑ በመለየት እና ነጥለን ለመድረስ የሚረዳንን ፕሮግራም መጻፍ እንችላለን። ይኼ የተለያዩ ጠቀሜታዎች አሉት። የመጀመሪያው ከላይ እንደተጠቀሰው ጠቋሚ ወይም ኢንዴክሰር ለመስራት ሲጠቅም፤ ሁለተኛው ደግሞ ጽሁፎችን ማሳጠር (ሰመራይዝ) ማድረግ ብንፈልግ ይሄ ግባት በመሆን ያገለግላል። በቀጣዩ ክፍል የምንጽፈው ፕሮግራም የቃላትን ስፍራ በ አረፍተነገራት ጠቋሚነት መለየት የሚያስችል ነው። 